variadics/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+/// but it can also be used in `match`, `if let ...`, and `let ... else` expressions.
+///
+/// Although it may somtimes be possible to use `var_expr!` in place of this macro, doing so may
+/// cause confusing errors.
+///
+/// # Examples
+///
+/// ```rust
+/// use variadics::{var_expr, var_args, var_type};
+///
+/// fn my_fn(var_args!(a, b, c): var_type!(usize, &str, bool)) {
+///     println!("{} {} {}", a, b, c);
+/// }
+/// my_fn(var_expr!(12, "hello", false));
+/// ```
+///
+/// ```rust
+/// use variadics::{var_expr, var_args};
+///
+/// let val = var_expr!(true, Some("foo"), 2);
+/// if let var_args!(true, Some(item), 0..=3) = val {
+///     println!("{}", item);
+/// }
+/// else {
+///     unreachable!();
+/// }
+/// ```
+///
+/// ```rust
+/// # use variadics::{var_expr, var_args};
+/// match var_expr!(true, Some(100), 5) {
+///     var_args!(false, _, _) => unreachable!(),
+///     var_args!(true, None, _) => unreachable!(),
+///     var_args!(true, Some(0..=10), _) => unreachable!(),
+///     var_args!(true, Some(a), b) => println!("{} {}", a, b),
+/// }
+///
+#[macro_export]
+macro_rules! var_args {
+    () => ( () );
+    ($a:pat $(, $b:pat)* $(,)?)  => ( ($a, $crate::var_args!($($b),*)) );
+}
+
+/// Macro creating a variadic tuple type from a list of types.
+///
+/// `var_expr!` can be used to define simple types but will result in confusing errors for more
+/// complex types. Use this macro, `var_type!` instead.
+///
+/// # Examples
+///
+/// ```rust
+/// # use std::collections::HashMap;
+/// use variadics::{var_expr, var_type};
+///
+/// // A simple variadic type. Although `var_expr!` would work in this case, it cannot handle
+/// // more complex types i.e. ones with generics.
+/// let list: var_type!(i32, bool, String) = Default::default();
+///
+/// // A more complex type:
+/// let list: var_type!(
+///     &'static str,
+///     HashMap<i32, i32>,
+///     <std::vec::Vec<bool> as IntoIterator>::Item,
+/// ) = var_expr!("foo", HashMap::new(), false);
+/// ```
+///
+/// Unfortunately, expressions and types cannot be handled using the same macro due to the
+/// undefeated [bastion of the turbofish](https://github.com/rust-lang/rust/blob/7fd15f09008dd72f40d76a5bebb60e3991095a5f/src/test/ui/parser/bastion-of-the-turbofish.rs).
+#[macro_export]
+macro_rules! var_type {
+    () => ( () );
+    ($a:ty $(, $b:ty)* $(,)?)  => ( ($a, $crate::var_type!($($b),*)) );
+}
+
+/// This macro generates a basic variadic trait where each element must fulfill the trait clause.
+/// Currently of limited usefulness.
+///
+/// ```rust
+/// use variadics::{var_expr, variadic_trait};
+///
+/// variadic_trait! {
+///     /// A variadic list of `Debug` items.
+///     pub variadic<T> DebugList where T: std::fmt::Debug {}
+/// }
+///
+/// let x = &var_expr!(1, "hello", 5.6);
+/// let _: &dyn DebugList = x;
+/// println!("{:?}", x);
+/// ```
+#[macro_export]
+macro_rules! variadic_trait {
```

Comment:
This file is the only change worth reviewing at all. Code below this review comment is all old, pretty much